### ISSUE TITLE
Custom components with text

### ISSUE THREAD

**Issue by epalese** [facebook/draft-js#353]
Hello,

I am trying to implement custom block components that include text as part of their content and what I want to achieve is a seamless integration of the text component in the custom component with the containing editor. 

I've modified the two following files in TeX example:
[TeXEditorExample](https://github.com/epalese/draft-js/blob/master/examples/tex/js/components/TeXEditorExample.js)
[TeXBlock](https://github.com/epalese/draft-js/blob/master/examples/tex/js/components/TeXBlock.js)

My first question is:
1. In the custom component I use an instance of Editor to handle the text. Is this a good approach? In the documentation a DraftEditorBlock is suggested but I haven't been able to make it work (set up its content). 

What I've done so far is handling the switch between the two editors (the main one and the one that lives in the custom block). The main problem was in the fact that only one Editor can be editable at time and the other one must have its readOnly property set to false.
1. Is there a way to handle the smooth transition between the two editors using cursor arrow keys? Now the cursor simply skip the custom block and moves to the next text block after it.
2. When I click on the Editor in the custom component, the cursor appears in the editor but it is on the position it was before (or at the beginning of the block if none). Is there a way to force the cursor to move just beneath the mouse pointer?

I am new with React and Draft-js and I think it's really awesome!

Any help with this would be great. Thanks.
Emanuele

**Comment by hellendag** [comment]
Nesting `Editor` instances within custom blocks is theoretically doable, but you will definitely need to set `readOnly` on the outer editor. Without this, selection handling is almost certain to break.

@bkniffler has been experimenting with this and may be able to lend a hand.

**Comment by bkniffler** [comment]
Hi @epalese, these are some great questions!

I think that using a nested draft editor is a valid approach, and I've gotten very smooth results (with the table and the captions here http://draft-wysiwyg.herokuapp.com/).

I saw that you use onClick, which leads to your second issue; draft-js uses onMouseDown for selection (I guess) and a click only results in the editor getting focused, with the selection being defaulted (beginning of the editor). If you use onMouseDown instead this will work as you'd expect it, give it a try.

The first one is very interesting and a bit tricky. You will need to keep track of what block will be focused next after arrow up/down/left/right key events. Draft-Js exposes onDownArrow/onUpArrow which you could use for a start. You will then need something like 

```
editorState.getCurrentContent().getBlockBefore(selection.getAnchorKey())
editorState.getCurrentContent().getBlockAfter(selection.getAnchorKey());
```

to get the next/previous block and do the readOnly switcheroo if the blocks type equals to 'atomic'. You will likely face a few more hickups, but I imagine it could work out very well. Keep us updated of you results 👍

**Comment by epalese** [comment]
Hey guys,

thanks for your promptly replies. 

@bkniffler I've played very quickly with onMouseDown and I had problems to control the propagation (it is both intercepted by both the inner and outer editor even if I explicitly call stopProgation()).
But I am going to spend some more time on it and on the suggestions you gave me and I'll keep you posted on what I manage to achieve.

I've seen draft-wysiwyg. I think you've done an amazing job with it!

Thanks

**Comment by clwilliams** [comment]
Hi, I was wondering if any progress had been made with getting a draft js editor working in a custom block, as this is exactly what I'm trying to do now . . . with difficultly! I'm relatively new to React / Redux / Draft js / this level of JS! I've got a popup to appear on my custom block to allow editing of the text, but doesn't feel as good to use as leaving and editing inline. I also can't get the tex example working as I hoped that might shed a little light on my problem, I followed instructions from https://github.com/facebook/draft-js#examples but I get errors in the console about missing files when I try to open the example tex index.html file? Any help or pointers appreciated! Many thanks, Claire
